### DISCUSSION TITLE
Issues trying to set states of other fields when <Controller> textfield is selected

### DISCUSSION THREAD

**Discussion by kaduppg** [react-hook-form/react-hook-form#5407] [Q&A] [answered]
Hey guys, just would like to see with you what is the problem in the following situation: 
Basically I'm starting out with react, so I dont have that much knowledge about it. I've been creating a few forms using react hook form v7 with material-ui, more specific im using a Stepper component from MUI, and right now im stuck!  The form is ok the validation is ok but I've got one TextField using the google maps api, so the idea is select the address from the component and set the other fields like (city, state and zip). I do can set the state value with the selected address for example, but when i do this the state field gets weird and i cannot type anything and when i try to submit again it pops up with the same required error even tho there's information in it. Plus the TextField which has been used as the google api doesn't get updated after select the item from it.  I'll paste the code here from the main stepper and the Step1 then you can see. I reckon there's something to do with the <Controller> .... Might be something stupid but i dont know which is the best way to do solve this problem.  

Cheers.

MAIN STEPPER  ----------------------------

function getStepContent(stepIndex) {
  switch (stepIndex) {
    case 0:
      return <Step1/>;
    case 1:
      return <Step2/>;
    case 2:
      return <Step3/>;
    case 3:
      return <Step4/>;
    case 4:
      return <Step5/>;
    default:
      return 'Unknown stepIndex';
  }
}

const defaultValues = {
  fisrtname:"",
  title:"",
  name:"",
  field:"",
  address:"",
  city:"",
  state:"",
  zip:"",
  whatsapp: "",
  remote: ""
}

const schema = Yup.object().shape({
  name:   Yup.string().required(),
  title:  Yup.string().required(),
  field:  Yup.string().required(),
  address: Yup.string().required(),
  city:   Yup.string().required(),
  state:  Yup.string().required(),
  zip:    Yup.string().required(),
  whatsapp: Yup.string().required(),
});

export default function MainStepper() {

  const classes = useStyles();
  const [activeStep, setActiveStep] = useState(0);
  const steps = getSteps();

  //const currentSchemaValidation = validationSchema[activeStep];

  const methods = useForm({
    defaultValues,
    resolver: yupResolver(schema),
    mode: "onChange"
  })

  const { handleSubmit, reset, trigger } = methods;

  const onSubmit = (data) => {
    alert(JSON.stringify(data));
    //handleNext();
  }

  const handleNext = async() => {
    const isValid = await trigger();
    if(isValid){
      setActiveStep((prevActiveStep)=> prevActiveStep + 1);
    }
  };

  const handleBack = () => {
    setActiveStep( (prevActiveStep) => prevActiveStep - 1);
  };

  const handleReset = () => {
    setActiveStep(0);
    reset();
  };

  return (
    <main className={classes.layout}>
        <Paper className={ classes.paper }>

            <Typography component="h1" variant="h4" align="center">Lawi</Typography>
            
            <Stepper activeStep={activeStep} alternativeLabel>
                {steps.map((step) => (
                  <Step key={step}>
                      <StepLabel>{step}</StepLabel>
                  </Step>
                ))}
            </Stepper>
            
            <div>
                {activeStep === steps.length ? (
                <div>
                    <Typography className={classes.instructions}>All steps completed</Typography>
                    <Button onClick={handleReset}>Reset</Button>
                </div>
                ) : (
                    <FormProvider {...methods} >
                     <form onSubmit={ handleSubmit(onSubmit)}>
                        <div>
                          <Fragment>
                          
                              {getStepContent(activeStep)}
                            
                              <div className={classes.buttons}>
                                  <Button 
                                      disabled={activeStep === 0} 
                                      onClick={handleBack} 
                                      className={classes.backButton} >
                                      Back
                                  </Button>

                                  {activeStep === steps.length -1 ? (
                                    <Button 
                                        type="submit" 
                                        variant="contained" 
                                        color="primary" 
                                     >
                                      Finalizar Cadastro
                                    </Button>
                                  ) : (
                                    <Button 
                                      variant="contained" 
                                      color="primary" 
                                      onClick={handleNext} >
                                      Next
                                    </Button>
                                  )}

                              </div>
                          
                          </Fragment>
                        </div>
                      </form>
                    </FormProvider>

                )}

            </div>
      </Paper>
    </main>
  );
}

STEP1 ------------------------------------------------------------------------------------------------------------------

export default function Step1 () {

  const autocomplete = { current: null } ;
  const {control, formState:{errors}} = useFormContext();
  const [address, setAddress] = useState('');
  const [state, setState] = useState('');

  useEffect(() => {
    autocomplete.current = new google.maps.places.Autocomplete(
        document.getElementById('address'), 
        {
            types: ["address"], 
            componentRestrictions: {'country' : ['br']},
            fields: ['place_id', 'address_component', 'name','formatted_address']
        } 
    )

  
    autocomplete.current.addListener("place_changed", handlePlaceSelect);
    
    function handlePlaceSelect (){

       let addressObject = autocomplete.current.getPlace()
       let address = addressObject.address_components
  
       console.log(addressObject.formatted_address);

       for (let item of address) {
         console.log(item)

         switch (item.types[0]) {

           case 'sublocality_level_1': // sub area
             console.log('bairro - '+item.long_name)
             break;
   
           case 'administrative_area_level_2':  // city
             console.log('cidade - '+item.long_name)
             break;
           case 'administrative_area_level_1': // state
             setState(item.short_name);
             console.log('estado - '+item.short_name)
             break;
           case 'postal_code':   //zip
             console.log('cep - '+item.short_name)
             break;
   
           default:
             break;
         }
       }
   
    }
  }, [])

    return (

      <Fragment>

        <Typography  variant="h5" gutterBottom>
          Sobre você
        </Typography>
      
        <Grid container spacing={3} >

          <Grid item xs={12} sm={3}>
    

            <Controller
                control={control}
                name="title"
                /* rules={{ required: "this field is required." }} */
                render={({ field }) => (
                  <TextField
                    id="title"
                    label="Título"
                    fullWidth
                    margin="normal"
                    {...field}
                    error={Boolean(errors?.title)}
                    helperText={errors.title?.message} 
                    select >

                    <MenuItem>selecione</MenuItem>
                    <MenuItem value="Dr.">Dr.</MenuItem>
                    <MenuItem value="Dra.">Dra.</MenuItem>

                  </TextField>
                )}
              />
          </Grid>

          <Grid item xs={12} sm={9}>
            <Controller
                control={control}
                name="name"
                render={({ field }) => (
                  <TextField
                    id="name"
                    label="Nome completo"
                    placeholder="Digite seu nome completo"
                    fullWidth
                    margin="normal"
                    {...field}
                    error={Boolean(errors?.name)}
                    helperText={errors.name?.message}
                  />
                )}
              />
          </Grid>

          <Grid item xs={12}>
            <Controller
              control={control}
              name="field"
              render={({field}) => (
                <TextField
                  id="field"
                  label="Principal área de atuação"
                  select
                  fullWidth
                  margin="normal"
                  {...field}
                  error={Boolean(errors?.field)}
                  helperText={errors.field?.message}>
                    <MenuItem>selecione</MenuItem>
                    <MenuItem value="1">Trabalhista</MenuItem>
                    <MenuItem value="2">Previdenciário</MenuItem>
                    <MenuItem value="3">Família</MenuItem>
                    <MenuItem value="4">Tributarista</MenuItem>
                    <MenuItem value="5">Criminal</MenuItem>
                </TextField>
              )}
            />
          </Grid>

        </Grid>

        <Box m={4} />
        
        <Typography  variant="h6" gutterBottom>
          Qual o endereco do seu escritorio
        </Typography>
         
        <Grid container spacing={3}>
          <Grid item xs={12}>
        
          <Controller
                control={control}
                name="address"
                render={({field})=>(
                  <TextField
                    id="address"
                    label="Endereço"
                    fullWidth
                    {...field}
                    value={address}
                    error={Boolean(errors?.address)}
                    helperText={errors.address?.message}
                  />
                )}
          />
                
             
          </Grid>
          <Grid item xs={12} sm={6}>
            <Controller
                control={control}
                name="city"
                render={({field})=>(
                  <TextField
                    id="city"
                    label="Cidade"
                    fullWidth
                    {...field}
                    error={Boolean(errors?.city)}
                    helperText={errors.city?.message}
                  /> 
                )}
              />
          </Grid>
          <Grid item xs={12} sm={3}>
             <Controller
                  control={control}
                  name="state"
                  render={({field})=>(
                    <TextField
                      id="state"
                      label="Estado"
                      fullWidth
                      {...field}
                      value={state} 
                      error={Boolean(errors?.state)}
                      helperText={errors.state?.message}
                    /> 
                  )}
            /> 
          </Grid>
          <Grid item xs={12} sm={3}>
            <Controller
                    control={control}
                    name="zip"
                    render={({field, value})=>(
                      <TextField
                        id="zip"
                        label="CEP"
                        fullWidth
                        {...field}
                        error={Boolean(errors?.zip)}
                        helperText={errors.zip?.message}
                      /> 
                    )}
              />
          </Grid>
          <Grid item xs={12} >
            <Controller
                control={control}
                name="whatsapp"
                render={({field, value})=>(
                  <TextField
                    id="whatsapp"
                    label="Seu numero no WhatsApp"
                    fullWidth
                    {...field}
                    error={Boolean(errors?.zip)}
                    helperText={errors.zip?.message}
                  /> 
                )}
            />
          </Grid>
          <Grid item xs={12} >
            <FormControlLabel
              control={<Checkbox color="secondary" name="remote" value="" />}
              label="Faz atendimento remoto?"
            />
          </Grid>
        </Grid>
        
      </Fragment>

    );
};

**Comment by wdfinch** [comment]
Please attach a code sandbox that just focuses just on the issue and I can take a look.

**Reply by kaduppg** [reply]
https://codesandbox.io/s/mystifying-proskuriakova-f077m?file=/src/App.js 

There it is! 
Again the problem is just when I select the address from google api and try to set the value for state, every thing works fine but the address doenst get updated, the state stop working and if i try to click the next button to validate again even tho state has an input the field still with an error.

**Comment by thanh-nguyen-95** [comment, marked as answer]
Changing the Address `<TextField />` like so works for me:

```jsx
<TextField
  id="address"
  label="Address"
  fullWidth
  {...field}
  inputRef={field.ref} // passing ref here, remove the 'value' prop (already included in field)
  error={Boolean(errors?.address)}
  helperText={errors.address?.message}
/>
```

**Reply by kaduppg** [reply]
For me nothing has changed =(, everything still the same. Same problems

**Reply by thanh-nguyen-95** [reply]
Before changing the `TextField`, I cannot type in the address input. However, after changing it like above, I can. Isn't it the problem you're facing?

If that's not the problem, please provide the steps to produce the problem.

**Reply by kaduppg** [reply]
Actually you can type without any problem in address field! The thing is ... when you enter with 1 or more letters the google api is gonna be called  and it returns the predictions. When you select one item it supposed to set the value inside the address TextField plus the state TextField (this one is happening, but after that i can't type anything into this field anymore), and when i press the next button the validation runs again and pops the error for State TextField even tho there is value.

**Reply by thanh-nguyen-95** [reply]
https://codesandbox.io/s/amazing-lewin-uwtk5?file=/src/Step1.js

This is a working version of your form. Try typing 'Rua' in your old version, the address field is indeed not working (just because it shows suggestion doesn't mean it's working correctly).

Key take-aways:

- `autocomplete` should be a `useRef` so it won't change every render
- Pass `inputRef={field.ref}` to each input to update `react-hook-form` values
- Use `setValue` to update `react-hook-form` values when user select an address

**Reply by bluebill1049** [reply]
thank you @thanh-nguyen-95 for the help. 🙏

**Reply by kaduppg** [reply]
Thanks a lot @thanh-nguyen-95, appreciate it. Didn't know about the useRef and the setValue. Now question... I was thinking to create a DataContext to storage all the data throughout the forms. Do I need to create one or do you think setValues is gonna be the same at the end ?

**Reply by thanh-nguyen-95** [reply]
Form values should all stay (if registered before handle) when you submit the form.

Try completing all steps and submit to see if the data persists. If not, you might need to persist the data somewhere else.

**Comment by kaduppg** [comment]
hey @thanh-nguyen-95, I have come across with another problem, same project https://codesandbox.io/s/aged-pond-i1vcn?file=/src/Step1.js, 
why this doesn't work (  onChange={() => field.onChange(handleFieldChange)} ). I have been trying to call the onChange but i,ve got the issue : Material-UI: A component is changing the controlled value state of Select to be uncontrolled, and on the react hook form says i have to use like this. So Im kinda confused now.

**Reply by thanh-nguyen-95** [reply]
I will not help you this time. It looks to me that you have no idea what you're doing at all.

My sugesstion is to learn (or re-learn) the basic of Javascript and React first, then apply it to a couple side projects with out any external libraries, just pure React. After all that, come back to fix the problems you're facing.

Here are some good resources that I used when I first learned React:

- Javascript: 
  - https://javascript.info/
  - https://eloquentjavascript.net/
  - https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/README.md
 
- React:
  - https://reactjs.org/docs/hello-world.html

You might think it's a waste of time to read all those boring books/documents but it will save much more when you're actually working on a project. Good luck!
